### ISSUE TITLE
Send Sluggish

### ISSUE THREAD

**Issue by numediatechnologies** [tin-cat/emailqueue#10]
Hi 

Thanx for the code, I improved it and added "to_name" and also changed mktime() to time() as that results is some warnings , time() more efficient for timestamp 'now', my research suggests.

Ok I have 30 000 emails queued and it is sending at about 3/sec, am using SendGrid API Free with 40 000 emails free to test, by this rate I will finish sending in 25 hours, now I have a list of 120 000 subscribers, this would take I think3-4 days, how to speedy things up?

Delivery:

I am calling the delivery.php directly via wget from cron job (not secure), my server wont fire the bash file, is the bash script file calling better that calling the delivery.php directly? If so why?

**My setting are the default:**

EnvironmentProduction
Maximum delivery timeout50 seconds
Delivery interval100 ms.
Maximum retry attempts3
Maximum delivers per call1,500
SMTP serversmtp.sendgrid.net

**Comment by roselan** [comment]
The main reason sendgrid throttles your emails is because your sender/domain is "greylisted". You can check your reputation with MxToolbox (there is some stuff in sendgrid too). Open a ticket with sendgrid if you want to know why, their support is top notch even with their free plan.

**Comment by tin-cat** [comment]
Hi @numediatechnologies , it's also well known that some smtp service providers would limit your delivery rate if you're using some kind of trial account, check with sendgrid to see if they apply delivery rate limits to free accounts.

Emailqueue tries to stablish a persistent connection to the SMTP server by using the "SMTPKeepAlive" parameter of the PHPMailer library. As long as the persistent connection is stablished (no way to check with PHPMailer), this should discard emailqueue as the bottleneck.

Also, as @roselan suggested, check out if your sender IP is blacklisted/greylisted somewhere.

Let us know your solution once found!

Thanks both!

**Comment by tin-cat** [comment]
Additionally, SMTP services like Amazon SES and many others limit your delivery rate based in their trust in you. For example,  If I remember right, Amazon SES starts you with a low delivery rate and raises it automatically over time if your bounce/spam rate is low enough.

meow.

**Comment by tin-cat** [comment]
BTW, @numediatechnologies, I do not recommend at all to call the delivery.php via wget. This means an instance of your web server is being used to deliver the emails, and this leads to lots of serious problems (note I said "leads", not "may lead"!)

For instance, your web server sure has some kind of maximum execution time limits, memory limits ... when calling delivery.php with a web request, this limits apply and might end the delivery execution unexpectedly just after a few seconds, or give other kind of errors related to web-serving.

It's strongly recommended that you execute the provided script, which is actually really simple: it just executes deliver.php vie the PHP CLI.

It might not work straightaway in your system because the "delivery" script has some hardcoded paths, just edit it to suit your environment and it should work ok.

Cheers!

**Comment by numediatechnologies** [comment]
Thank you guys.

**Increasing Send Rate of the Delivery Script:**
@roselan It is not greylisted its a new free account on their shared IP.
@tin-cat I cant access their support as fast as I wish on this free trial account.

**Solution and problem:**
SendGrid allows one to submit up to 10 000 recipients per time using their  SMTP-API custom header, https://sendgrid.com/docs/API_Reference/SMTP_API/using_the_smtp_api.html , they recommend 1000 though. 

Problem arising is that each body of the email is personalised and has different content each week, so there is no way to also submit the body with each connection using this SMSTP-API of Sangrid, which means, I will have to modify the delivery php file to not include the body when sending to that 1000 batch per sending request, I have not investigated, but I presume, SendGrid possibly has templating system of the message body, so that I include that one body and SendGrid replaces certain variables for personalisation, including the unsubscribe URL.

I will investigate this further.

I will also investigate using **"BCC"** but in my ad hoc analysis now that option fails for the body reason.

Last complication is, with these methods , how do I include the List-Unsubscribe custom header that @tin-cat rightfully includes.


My Server refuses to fire my bash cron job only wget http it fires which causes problems as @tin-cat mentioned, maybe something to do with the server or how I setup the cron?

*	*	*	*	*	/[ROOT_DIRECTORY]/emailqueue/scripts/delivery.sh

**Comment by tin-cat** [comment]
Hi @numediatechnologies, I haven't dived deep into the sendgrid documentation you linked, but it seems you're trying to use some kind of pattern-based email sending service that just happens to have an API implemented via SMTP (if I understood well).

If that's the case, please note EmailQueue is not built to work with that kind of non-standard SMTP services, it is just built to send normal emails via standard SMTP requests. This means you have to provide the complete PlainText/HTML body of your message everytime you send an email, and you'll have to stick to the standard headers provided by the EmailQueue functionalities (like the unsubscribe URL header you mention), or provide your own via the additional headers capabilities.

My recommendation if you still want to work with Emailqueue is to use just Sendgrid's regular SMTP Relay service, not their advanced API, which is definitely aimed to more specific environments and provides lots of advanced functionalities outside of the scope of EmailQueue.

Cheers!

**Comment by numediatechnologies** [comment]
@tin-cat Thank you for your assistance, you certainly make sense.

However, I was using SMTP so am basically trying to solve the problem of the script being too slow for my intended use, eg send to 100 000 email list will take many days  at 1 email/3-4 sec. I have used Amazon SES before and they allow 50 emails/sec , the  delivery system worked well if I used BCC field, so am trying to find a way to increase the delivery rate.
